File: julia/specialized/FiniteGrobner.jl
module FiniteGrobner
    export Term, Polynomial, TermOrder, set_term_order, set_modulus, grobner_basis, s_polynomial, reduce_polynomial

    mutable struct Term
        coefficient::UInt64
        exponents::Vector{Int}
    end

    @enum TermOrder Lex GrLex RevLex

    const TERM_ORDER = Ref(Lex)
    const MODULUS = Ref{UInt64}(7)

    function set_term_order(order::TermOrder)
        TERM_ORDER[] = order
    end

    function set_modulus(m::UInt64)
        MODULUS[] = m
    end

    function get_modulus()
        return MODULUS[]
    end

    function mod_inverse(a::UInt64, m::UInt64)
        a_i64 = Int64(a)
        m_i64 = Int64(m)
        m0 = m_i64
        y = 0
        x = 1

        if m == 1
            return UInt64(0)
        end

        while a_i64 > 1
            q = div(a_i64, m_i64)
            t = m_i64

            m_i64 = a_i64 % m_i64
            a_i64 = t
            t = y

            y = x - q * y
            x = t
        end

        if x < 0
            x = x + m0
        end

        return UInt64(x % m0)
    end

    function compare(t1::Term, t2::Term)
        order = TERM_ORDER[]
        if order == Lex
            return cmp(t1.exponents, t2.exponents)
        elseif order == GrLex
            d1 = sum(t1.exponents)
            d2 = sum(t2.exponents)
            if d1 != d2
                return cmp(d1, d2)
            else
                return cmp(t1.exponents, t2.exponents)
            end
        elseif order == RevLex
            d1 = sum(t1.exponents)
            d2 = sum(t2.exponents)
            if d1 != d2
                return cmp(d1, d2)
            else
                return cmp(reverse(t1.exponents), reverse(t2.exponents))
            end
        end
    end

    mutable struct Polynomial
        terms::Vector{Term}
        function Polynomial(terms::Vector{Term})
            sort!(terms, by=t -> t, lt= (a,b) -> compare(a,b) == 1)
            filter!(t -> t.coefficient != 0, terms)
            new(terms)
        end
    end

    import Base: +, -, *

    function +(p1::Polynomial, p2::Polynomial)
        result = deepcopy(p1.terms)
        m = MODULUS[]
        for term in p2.terms
            found = false
            for res_term in result
                if res_term.exponents == term.exponents
                    res_term.coefficient = (res_term.coefficient + term.coefficient) % m
                    found = true
                    break
                end
            end
            if !found
                push!(result, deepcopy(term))
            end
        end
        Polynomial(result)
    end

    function -(p1::Polynomial, p2::Polynomial)
        m = MODULUS[]
        neg_p2_terms = [Term((m - (term.coefficient % m)) % m, term.exponents) for term in p2.terms]
        p1 + Polynomial(neg_p2_terms)
    end

    function *(p1::Polynomial, p2::Polynomial)
        new_terms = Term[]
        m = MODULUS[]
        for t1 in p1.terms
            for t2 in p2.terms
                coeff = (t1.coefficient * t2.coefficient) % m
                exps = [e1 + e2 for (e1, e2) in zip(t1.exponents, t2.exponents)]
                push!(new_terms, Term(coeff, exps))
            end
        end
        # combine like terms
        combined_terms = Term[]
        for term in new_terms
            found = false
            for c_term in combined_terms
                if c_term.exponents == term.exponents
                    c_term.coefficient = (c_term.coefficient + term.coefficient) % m
                    found = true
                    break
                end
            end
            if !found
                push!(combined_terms, term)
            end
        end
        Polynomial(combined_terms)
    end

    function leading_term(p::Polynomial)
        if isempty(p.terms)
            return Term(UInt64(0), [])
        end
        return p.terms[1]
    end

    function s_polynomial(p1::Polynomial, p2::Polynomial)
        lt1 = leading_term(p1)
        lt2 = leading_term(p2)

        if isempty(p1.terms) || isempty(p2.terms)
            return Polynomial(Term[])
        end

        lcm_exponents = [max(e1, e2) for (e1, e2) in zip(lt1.exponents, lt2.exponents)]

        s1_exp = [l - e for (l, e) in zip(lcm_exponents, lt1.exponents)]
        s2_exp = [l - e for (l, e) in zip(lcm_exponents, lt2.exponents)]

        m = MODULUS[]
        inv1 = mod_inverse(lt1.coefficient, m)
        inv2 = mod_inverse(lt2.coefficient, m)

        s1_term = Term(lt2.coefficient, s1_exp)
        s2_term = Term(lt1.coefficient, s2_exp)

        term_poly1 = Polynomial([s1_term])
        term_poly2 = Polynomial([s2_term])

        return (p1 * term_poly1) - (p2 * term_poly2)
    end

    function make_monic(p::Polynomial)
        if isempty(p.terms)
            return p
        end
        m = MODULUS[]
        lt_coeff = leading_term(p).coefficient
        inv_coeff = mod_inverse(lt_coeff, m)
        new_terms = [Term((t.coefficient * inv_coeff) % m, t.exponents) for t in p.terms]
        return Polynomial(new_terms)
    end

    function reduce_polynomial(p::Polynomial, basis::Vector{Polynomial})
        poly = deepcopy(p)
        m = MODULUS[]
        remainder = Term[]

        while !isempty(poly.terms)
            reduced = false
            lt_poly = leading_term(poly)

            for b in basis
                if isempty(b.terms)
                    continue
                end
                lt_b = leading_term(b)
                if all(lt_poly.exponents .>= lt_b.exponents)
                    inv_b = mod_inverse(lt_b.coefficient, m)
                    scale_coeff = (lt_poly.coefficient * inv_b) % m
                    scale_exps = [e1 - e2 for (e1, e2) in zip(lt_poly.exponents, lt_b.exponents)]

                    scale_term = Term(scale_coeff, scale_exps)
                    scaled_b = b * Polynomial([scale_term])
                    poly = poly - scaled_b
                    reduced = true
                    break
                end
            end

            if !reduced
                push!(remainder, deepcopy(leading_term(poly)))
                popfirst!(poly.terms)
            end
        end
        return Polynomial(remainder)
    end

    function grobner_basis(polynomials::Vector{Polynomial})
        basis = deepcopy(polynomials)
        n_orig = length(basis)
        pair_set = [(i, j) for i in 1:length(basis) for j in (i+1):length(basis)]

        while !isempty(pair_set)
            (i, j) = popfirst!(pair_set)
            s_poly = s_polynomial(basis[i], basis[j])
            h = reduce_polynomial(s_poly, basis)

            if !isempty(h.terms)
                push!(basis, h)
                n = length(basis)
                for k in 1:(n-1)
                    push!(pair_set, (k, n))
                end
            end
        end

        reduced_basis = Polynomial[]
        for i in 1:length(basis)
            basis_excluding_self = vcat(basis[1:i-1], basis[i+1:end])
            reduced = reduce_polynomial(basis[i], basis_excluding_self)
            if !isempty(reduced.terms)
                push!(reduced_basis, make_monic(reduced))
            end
        end

        return reduced_basis
    end

    function main(n::Int)
        set_modulus(UInt64(7))
        polynomials = Polynomial[]
        if n == 4
            # cyclic 4
            p1 = Polynomial([
                Term(UInt64(1), [1, 0, 0, 0]),
                Term(UInt64(1), [0, 1, 0, 0]),
                Term(UInt64(1), [0, 0, 1, 0]),
                Term(UInt64(1), [0, 0, 0, 1]),
            ])
            p2 = Polynomial([
                Term(UInt64(1), [1, 1, 0, 0]),
                Term(UInt64(1), [0, 1, 1, 0]),
                Term(UInt64(1), [0, 0, 1, 1]),
                Term(UInt64(1), [1, 0, 0, 1]),
            ])
            p3 = Polynomial([
                Term(UInt64(1), [1, 1, 1, 0]),
                Term(UInt64(1), [0, 1, 1, 1]),
                Term(UInt64(1), [1, 0, 1, 1]),
                Term(UInt64(1), [1, 1, 0, 1]),
            ])
            p4 = Polynomial([
                Term(UInt64(1), [1, 1, 1, 1]),
                Term(UInt64(get_modulus() - 1), [0, 0, 0, 0]),
            ])
            polynomials = [p1, p2, p3, p4]
        end
        if n == 5
            # cyclic 5
            p1 = Polynomial([
                Term(UInt64(1), [1, 0, 0, 0, 0]),
                Term(UInt64(1), [0, 1, 0, 0, 0]),
                Term(UInt64(1), [0, 0, 1, 0, 0]),
                Term(UInt64(1), [0, 0, 0, 1, 0]),
                Term(UInt64(1), [0, 0, 0, 0, 1]),
            ])
            p2 = Polynomial([
                Term(UInt64(1), [1, 1, 0, 0, 0]),
                Term(UInt64(1), [0, 1, 1, 0, 0]),
                Term(UInt64(1), [0, 0, 1, 1, 0]),
                Term(UInt64(1), [0, 0, 0, 1, 1]),
                Term(UInt64(1), [1, 0, 0, 0, 1]),
            ])
            p3 = Polynomial([
                Term(UInt64(1), [1, 1, 1, 0, 0]),
                Term(UInt64(1), [0, 1, 1, 1, 0]),
                Term(UInt64(1), [0, 0, 1, 1, 1]),
                Term(UInt64(1), [1, 0, 0, 1, 1]),
                Term(UInt64(1), [1, 1, 0, 0, 1]),
            ])
            p4 = Polynomial([
                Term(UInt64(1), [1, 1, 1, 1, 0]),
                Term(UInt64(1), [0, 1, 1, 1, 1]),
                Term(UInt64(1), [1, 0, 1, 1, 1]),
                Term(UInt64(1), [1, 1, 0, 1, 1]),
                Term(UInt64(1), [1, 1, 1, 0, 1]),
            ])
            p5 = Polynomial([
                Term(UInt64(1), [1, 1, 1, 1, 1]),
                Term(UInt64(get_modulus() - 1), [0, 0, 0, 0, 0]),
            ])

            polynomials = [p1, p2, p3, p4, p5]
        end
        if n == 6
            # cyclic 6
            p1 = Polynomial([
                Term(UInt64(1), [1, 0, 0, 0, 0, 0]),
                Term(UInt64(1), [0, 1, 0, 0, 0, 0]),
                Term(UInt64(1), [0, 0, 1, 0, 0, 0]),
                Term(UInt64(1), [0, 0, 0, 1, 0, 0]),
                Term(UInt64(1), [0, 0, 0, 0, 1, 0]),
                Term(UInt64(1), [0, 0, 0, 0, 0, 1]),
            ])
            p2 = Polynomial([
                Term(UInt64(1), [1, 1, 0, 0, 0, 0]),
                Term(UInt64(1), [0, 1, 1, 0, 0, 0]),
                Term(UInt64(1), [0, 0, 1, 1, 0, 0]),
                Term(UInt64(1), [0, 0, 0, 1, 1, 0]),
                Term(UInt64(1), [0, 0, 0, 0, 1, 1]),
                Term(UInt64(1), [1, 0, 0, 0, 0, 1]),
            ])
            p3 = Polynomial([
                Term(UInt64(1), [1, 1, 1, 0, 0, 0]),
                Term(UInt64(1), [0, 1, 1, 1, 0, 0]),
                Term(UInt64(1), [0, 0, 1, 1, 1, 0]),
                Term(UInt64(1), [0, 0, 0, 1, 1, 1]),
                Term(UInt64(1), [1, 0, 0, 0, 1, 1]),
                Term(UInt64(1), [1, 1, 0, 0, 0, 1]),
            ])
            p4 = Polynomial([
                Term(UInt64(1), [1, 1, 1, 1, 0, 0]),
                Term(UInt64(1), [0, 1, 1, 1, 1, 0]),
                Term(UInt64(1), [0, 0, 1, 1, 1, 1]),
                Term(UInt64(1), [1, 0, 0, 1, 1, 1]),
                Term(UInt64(1), [1, 1, 0, 0, 1, 1]),
                Term(UInt64(1), [1, 1, 1, 0, 0, 1]),
            ])
            p5 = Polynomial([
                Term(UInt64(1), [1, 1, 1, 1, 1, 0]),
                Term(UInt64(1), [0, 1, 1, 1, 1, 1]),
                Term(UInt64(1), [1, 0, 1, 1, 1, 1]),
                Term(UInt64(1), [1, 1, 0, 1, 1, 1]),
                Term(UInt64(1), [1, 1, 1, 0, 1, 1]),
                Term(UInt64(1), [1, 1, 1, 1, 0, 1]),
            ])
            p6 = Polynomial([
                Term(UInt64(1), [1, 1, 1, 1, 1, 1]),
                Term(UInt64(get_modulus() - 1), [0, 0, 0, 0, 0, 0]),
            ])

            polynomials = [p1, p2, p3, p4, p5, p6]
        end
        println("Initial polynomials:")
        for p in polynomials
            println(p.terms)
        end

        set_term_order(Lex)
        basis = grobner_basis(polynomials)

        println("\nGrobner basis (Lex):")
        println("Basis size: ", length(basis))
        for b in basis
            println(b.terms)
        end
    end

end

if abspath(PROGRAM_FILE) == @__FILE__
    n = 4 # Default to cyclic 4
    if length(ARGS) > 0
        n = parse(Int, ARGS[1])
    end
    FiniteGrobner.main(n)
end

File: cpp/specialized/FiniteGrobner.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <chrono>
#include <unordered_set>

using namespace std;

enum TermOrder { Lex, GrLex, RevLex };
TermOrder TERM_ORDER = Lex;
uint64_t MODULUS = 7;

uint64_t mod_inverse(uint64_t a, uint64_t m) {
    if (m == 1) return 0;
    int64_t a_i64 = a;
    int64_t m_i64 = m;
    int64_t m0 = m_i64;
    int64_t y = 0, x = 1;
    while (a_i64 > 1) {
        int64_t q = a_i64 / m_i64;
        int64_t t = m_i64;
        m_i64 = a_i64 % m_i64;
        a_i64 = t;
        t = y;
        y = x - q * y;
        x = t;
    }
    if (x < 0) x += m0;
    return (uint64_t)(x % m0);
}

struct Term {
    uint64_t coefficient;
    vector<int> exponents;
    Term() : coefficient(0) {}
    Term(uint64_t c, vector<int> e) : coefficient(c), exponents(std::move(e)) {}
};

inline int compare_terms(const Term& t1, const Term& t2) {
    // Only Lex implemented as required
    for (size_t i = 0; i < t1.exponents.size() && i < t2.exponents.size(); ++i) {
        if (t1.exponents[i] < t2.exponents[i]) return -1;
        if (t1.exponents[i] > t2.exponents[i]) return 1;
    }
    if (t1.exponents.size() < t2.exponents.size()) return -1;
    if (t1.exponents.size() > t2.exponents.size()) return 1;
    return 0;
}

class Polynomial {
public:
    vector<Term> terms;
    Polynomial() {}
    Polynomial(vector<Term> _terms) {
        terms.reserve(_terms.size());
        for (auto& t : _terms) {
            if (t.coefficient != 0) {
                terms.push_back(std::move(t));
            }
        }
        sort(terms.begin(), terms.end(), [](const Term& a, const Term& b) {
            return compare_terms(a, b) > 0;
        });
    }
};

bool polynomials_equal(const Polynomial& p1, const Polynomial& p2) {
    if (p1.terms.size() != p2.terms.size()) return false;
    for (size_t i = 0; i < p1.terms.size(); i++) {
        if (p1.terms[i].coefficient != p2.terms[i].coefficient || p1.terms[i].exponents != p2.terms[i].exponents) {
            return false;
        }
    }
    return true;
}

inline bool operator==(const Polynomial& p1, const Polynomial& p2) {
    return polynomials_equal(p1, p2);
}

struct PolyHash {
    size_t operator()(const Polynomial& p) const {
        size_t hash = 0;
        for (const auto& t : p.terms) {
            hash ^= t.coefficient + 0x9e3779b9 + (hash << 6) + (hash >> 2);
            for (int e : t.exponents) {
                hash ^= e + 0x9e3779b9 + (hash << 6) + (hash >> 2);
            }
        }
        return hash;
    }
};

Polynomial add_poly(const Polynomial& p1, const Polynomial& p2) {
    vector<Term> result = p1.terms;
    for (const auto& term2 : p2.terms) {
        bool found = false;
        for (auto& res_term : result) {
            if (res_term.exponents == term2.exponents) {
                res_term.coefficient = (res_term.coefficient + term2.coefficient) % MODULUS;
                found = true;
                break;
            }
        }
        if (!found) result.push_back(term2);
    }
    return Polynomial(std::move(result));
}

Polynomial sub_poly(const Polynomial& p1, const Polynomial& p2) {
    vector<Term> result = p1.terms;
    for (const auto& term : p2.terms) {
        bool found = false;
        for (auto& res_term : result) {
            if (res_term.exponents == term.exponents) {
                res_term.coefficient = (MODULUS + res_term.coefficient - term.coefficient) % MODULUS;
                found = true;
                break;
            }
        }
        if (!found) {
            result.push_back(Term((MODULUS - (term.coefficient % MODULUS)) % MODULUS, term.exponents));
        }
    }
    return Polynomial(std::move(result));
}


Polynomial multiply_by_term(const Polynomial& p, const Term& term) {
    vector<Term> new_terms;
    new_terms.reserve(p.terms.size());
    for (const auto& t : p.terms) {
        vector<int> exps(t.exponents.size());
        for (size_t i = 0; i < exps.size(); ++i) {
            exps[i] = t.exponents[i] + term.exponents[i];
        }
        new_terms.push_back(Term((t.coefficient * term.coefficient) % MODULUS, std::move(exps)));
    }
    return Polynomial(std::move(new_terms));
}


Term leading_term(const Polynomial& p) {
    if (p.terms.empty()) return {0, vector<int>()};
    return p.terms[0];
}

Polynomial s_polynomial(const Polynomial& p1, const Polynomial& p2) {
    Term lt1 = leading_term(p1);
    Term lt2 = leading_term(p2);
    if (p1.terms.empty() || p2.terms.empty()) return Polynomial();

    vector<int> lcm_exps(lt1.exponents.size());
    vector<int> s1_exp(lt1.exponents.size());
    vector<int> s2_exp(lt2.exponents.size());
    for (size_t i = 0; i < lcm_exps.size(); ++i) {
        lcm_exps[i] = max(lt1.exponents[i], lt2.exponents[i]);
        s1_exp[i] = lcm_exps[i] - lt1.exponents[i];
        s2_exp[i] = lcm_exps[i] - lt2.exponents[i];
    }

    Term s1_term(lt2.coefficient, std::move(s1_exp));
    Term s2_term(lt1.coefficient, std::move(s2_exp));

    return sub_poly(multiply_by_term(p1, s1_term), multiply_by_term(p2, s2_term));
}

Polynomial make_monic(const Polynomial& p) {
    if (p.terms.empty()) return p;
    uint64_t lt_coeff = leading_term(p).coefficient;
    uint64_t inv_coeff = mod_inverse(lt_coeff, MODULUS);
    vector<Term> new_terms;
    for (const auto& t : p.terms) {
        new_terms.push_back(Term((t.coefficient * inv_coeff) % MODULUS, t.exponents));
    }
    return Polynomial(std::move(new_terms));
}

Polynomial reduce_polynomial(const Polynomial& p, const vector<Polynomial>& basis) {
    Polynomial poly = p;
    vector<Term> remainder;
    while (!poly.terms.empty()) {
        bool reduced = false;
        Term lt_poly = leading_term(poly);
        for (const auto& b : basis) {
            if (b.terms.empty()) continue;
            Term lt_b = leading_term(b);
            bool can_reduce = true;
            for (size_t i = 0; i < lt_poly.exponents.size(); ++i) {
                if (lt_poly.exponents[i] < lt_b.exponents[i]) {
                    can_reduce = false;
                    break;
                }
            }
            if (can_reduce) {
                uint64_t inv_b = mod_inverse(lt_b.coefficient, MODULUS);
                uint64_t scale_coeff = (lt_poly.coefficient * inv_b) % MODULUS;
                vector<int> scale_exps(lt_poly.exponents.size());
                for (size_t i = 0; i < scale_exps.size(); ++i) {
                    scale_exps[i] = lt_poly.exponents[i] - lt_b.exponents[i];
                }
                Polynomial scaled_b = multiply_by_term(b, {scale_coeff, scale_exps});
                poly = sub_poly(poly, scaled_b);
                reduced = true;
                break;
            }
        }
        if (!reduced) {
            remainder.push_back(std::move(poly.terms[0]));
            poly.terms.erase(poly.terms.begin());
        }
    }
    return Polynomial(std::move(remainder));
}

vector<Polynomial> grobner_basis(vector<Polynomial> basis) {
    unordered_set<Polynomial, PolyHash> basis_set;
    for (const auto& poly : basis) {
        basis_set.insert(poly);
    }

    vector<pair<int, int>> pair_set;
    for (size_t i = 0; i < basis.size(); i++) {
        for (size_t j = i + 1; j < basis.size(); j++) {
            pair_set.push_back({(int)i, (int)j});
        }
    }
    while (!pair_set.empty()) {
        auto p = pair_set.front();
        pair_set.erase(pair_set.begin());
        
        Polynomial s_poly = s_polynomial(basis[p.first], basis[p.second]);
        Polynomial h = reduce_polynomial(s_poly, basis);
        // h = make_monic(h);
        
        if (!h.terms.empty() && basis_set.insert(h).second) {
            int n = basis.size();
            basis.push_back(h);
            for (int k = 0; k < n; k++) {
                pair_set.push_back({k, n});
            }
        }
    }
    vector<Polynomial> reduced_basis;
    for (size_t i = 0; i < basis.size(); i++) {
        vector<Polynomial> basis_excluding_self;
        for (size_t j = 0; j < basis.size(); j++) {
            if (i != j) basis_excluding_self.push_back(basis[j]);
        }
        Polynomial reduced = reduce_polynomial(basis[i], basis_excluding_self);
        if (!reduced.terms.empty()) {
            reduced_basis.push_back(make_monic(reduced));
        }
    }
    return reduced_basis;
}

int main(int argc, char** argv) {
    int n = 4;
    if (argc > 1) {
        n = stoi(argv[1]);
    }
    
    MODULUS = 7;
    TERM_ORDER = Lex;
    vector<Polynomial> polynomials;

    if (n == 4) {
        cout << "C++ specialized vec exponent cyclic 4\n";
        polynomials = {
            Polynomial({ {1, {1, 0, 0, 0}}, {1, {0, 1, 0, 0}}, {1, {0, 0, 1, 0}}, {1, {0, 0, 0, 1}} }),
            Polynomial({ {1, {1, 1, 0, 0}}, {1, {0, 1, 1, 0}}, {1, {0, 0, 1, 1}}, {1, {1, 0, 0, 1}} }),
            Polynomial({ {1, {1, 1, 1, 0}}, {1, {0, 1, 1, 1}}, {1, {1, 0, 1, 1}}, {1, {1, 1, 0, 1}} }),
            Polynomial({ {1, {1, 1, 1, 1}}, {6, {0, 0, 0, 0}} })
        };
    } else if (n == 5) {
        cout << "C++ specialized vec exponent cyclic 5\n";
        polynomials = {
            Polynomial({ {1, {1, 0, 0, 0, 0}}, {1, {0, 1, 0, 0, 0}}, {1, {0, 0, 1, 0, 0}}, {1, {0, 0, 0, 1, 0}}, {1, {0, 0, 0, 0, 1}} }),
            Polynomial({ {1, {1, 1, 0, 0, 0}}, {1, {0, 1, 1, 0, 0}}, {1, {0, 0, 1, 1, 0}}, {1, {0, 0, 0, 1, 1}}, {1, {1, 0, 0, 0, 1}} }),
            Polynomial({ {1, {1, 1, 1, 0, 0}}, {1, {0, 1, 1, 1, 0}}, {1, {0, 0, 1, 1, 1}}, {1, {1, 0, 0, 1, 1}}, {1, {1, 1, 0, 0, 1}} }),
            Polynomial({ {1, {1, 1, 1, 1, 0}}, {1, {0, 1, 1, 1, 1}}, {1, {1, 0, 1, 1, 1}}, {1, {1, 1, 0, 1, 1}}, {1, {1, 1, 1, 0, 1}} }),
            Polynomial({ {1, {1, 1, 1, 1, 1}}, {6, {0, 0, 0, 0, 0}} })
        };
    } else if (n == 6) {
        cout << "C++ specialized vec exponent cyclic 6\n";
        polynomials = {
            Polynomial({ {1, {1, 0, 0, 0, 0, 0}}, {1, {0, 1, 0, 0, 0, 0}}, {1, {0, 0, 1, 0, 0, 0}}, {1, {0, 0, 0, 1, 0, 0}}, {1, {0, 0, 0, 0, 1, 0}}, {1, {0, 0, 0, 0, 0, 1}} }),
            Polynomial({ {1, {1, 1, 0, 0, 0, 0}}, {1, {0, 1, 1, 0, 0, 0}}, {1, {0, 0, 1, 1, 0, 0}}, {1, {0, 0, 0, 1, 1, 0}}, {1, {0, 0, 0, 0, 1, 1}}, {1, {1, 0, 0, 0, 0, 1}} }),
            Polynomial({ {1, {1, 1, 1, 0, 0, 0}}, {1, {0, 1, 1, 1, 0, 0}}, {1, {0, 0, 1, 1, 1, 0}}, {1, {0, 0, 0, 1, 1, 1}}, {1, {1, 0, 0, 0, 1, 1}}, {1, {1, 1, 0, 0, 0, 1}} }),
            Polynomial({ {1, {1, 1, 1, 1, 0, 0}}, {1, {0, 1, 1, 1, 1, 0}}, {1, {0, 0, 1, 1, 1, 1}}, {1, {1, 0, 0, 1, 1, 1}}, {1, {1, 1, 0, 0, 1, 1}}, {1, {1, 1, 1, 0, 0, 1}} }),
            Polynomial({ {1, {1, 1, 1, 1, 1, 0}}, {1, {0, 1, 1, 1, 1, 1}}, {1, {1, 0, 1, 1, 1, 1}}, {1, {1, 1, 0, 1, 1, 1}}, {1, {1, 1, 1, 0, 1, 1}}, {1, {1, 1, 1, 1, 0, 1}} }),
            Polynomial({ {1, {1, 1, 1, 1, 1, 1}}, {6, {0, 0, 0, 0, 0, 0}} })
        };
    }

    for (int i = 0; i < 10; ++i) {
        auto basis = grobner_basis(polynomials);
        cout << "Iteration " << i << " complete\n";
    }

    return 0;
}
